src/compiler/checker.ts

```diff
@@ -7515,6 +7527,54 @@ module ts {
             return flags & flagsToCheck;
         }
 
+        function checkDuplicateDeclarationOfSymbolInSameBlock(symbol: Symbol): void {
+            if ( !symbol.declarations) return;
+            
+            var declarations = symbol.declarations;
+            var duplicateDeclarationIndexes: Map<boolean> = {};
+            var signatures = getSignaturesOfSymbol(symbol);
+
+            for (var i = 0; i < signatures.length; i++) {
+                var signature1 = signatures[i];
+                for (var j = 0; j < i; j++) {

```

Comment:
So I have strong reservations about this. In `lib.d.ts` we have 122 overloads for `addEventListener`. As an `O(n^2)` operation, this is clearly expensive (14884 iterations).

Two things I could think of:
* If signature declarations could be totally ordered, one could "sort" the list of signatures, then pass through and find runs of duplicates (`O(n log n)` - or ~845 iterations for `addEventListener`)
* Find a way to serialize each signature declaration along with its parent, put each declaration into a map if it doesn't currently exist; if it does, report an error. (`O(n) assuming no background collisions, 122 iterations for `addEventListener`, but serialization could be slow).

@ahejl